doc/examples/qsv_transcode.c

```diff
@@ -0,0 +1,442 @@
+                             ofmt_ctx->streams[0]->time_base);
+        if ((ret = av_interleaved_write_frame(ofmt_ctx, enc_pkt)) < 0) {
+            fprintf(stderr, "Error during writing data to output file. "
+                    "Error code: %s\n", av_err2str(ret));
+            return ret;
+        }
+    }
+
+end:
+    if (ret == AVERROR_EOF)
+        return 0;
+    ret = ((ret == AVERROR(EAGAIN)) ? 0:-1);
+    return ret;
+}
+
+static int dec_enc(AVPacket *pkt, const AVCodec *enc_codec, char *optstr)
+{
+    AVFrame *frame;
+    int ret = 0;
+
+    ret = avcodec_send_packet(decoder_ctx, pkt);
+    if (ret < 0) {
+        fprintf(stderr, "Error during decoding. Error code: %s\n", av_err2str(ret));
+        return ret;
+    }
+
+    while (ret >= 0) {
+        if (!(frame = av_frame_alloc()))
+            return AVERROR(ENOMEM);
+
+        ret = avcodec_receive_frame(decoder_ctx, frame);
+        if (ret == AVERROR(EAGAIN) || ret == AVERROR_EOF) {
+            av_frame_free(&frame);
+            return 0;
+        } else if (ret < 0) {
+            fprintf(stderr, "Error while decoding. Error code: %s\n", av_err2str(ret));
+            goto fail;
+        }
+        if (!initialized) {
+            AVDictionaryEntry *e = NULL;
+            AVDictionary *opts = NULL;
+            AVStream *ost;
+            /* we need to ref hw_frames_ctx of decoder to initialize encoder's codec.
+               Only after we get a decoded frame, can we obtain its hw_frames_ctx */
+            encoder_ctx->hw_frames_ctx = av_buffer_ref(decoder_ctx->hw_frames_ctx);
+            if (!encoder_ctx->hw_frames_ctx) {
+                ret = AVERROR(ENOMEM);
+                goto fail;
+            }
+            /* set AVCodecContext Parameters for encoder, here we keep them stay
+             * the same as decoder.
+             */
+            encoder_ctx->time_base = av_inv_q(decoder_ctx->framerate);
+            encoder_ctx->pix_fmt   = AV_PIX_FMT_QSV;
+            encoder_ctx->width     = decoder_ctx->width;
+            encoder_ctx->height    = decoder_ctx->height;
+            if ((ret = str_to_dict(optstr, &opts)) < 0) {
+                fprintf(stderr, "Failed to set encoding parameter.\n");
+                goto fail;
+            }
+            /* There is no "framerate" option in commom option list. Use "-r" to
+            * set framerate, which is compatible with ffmpeg commandline. The
+            * video is assumed to be average frame rate, so set time_base to
+            * 1/framerate. */
+            e = av_dict_get(opts, "r", NULL, 0);
+            if (e) {
+                encoder_ctx->framerate = av_d2q(atof(e->value), INT_MAX);
+                encoder_ctx->time_base = av_inv_q(encoder_ctx->framerate);
+            }
+            if ((ret = avcodec_open2(encoder_ctx, enc_codec, &opts)) < 0) {
+                fprintf(stderr, "Failed to open encode codec. Error code: %s\n",
+                        av_err2str(ret));
```

Comment:
Add `av_dict_free(&opts);` to free opts.

Reply:
Done, thanks